core/objects.go

```diff
@@ -361,6 +397,57 @@ type Certificate struct {
 	Expires time.Time  `db:"expires"`
 }
 
+func (cert Certificate) MatchesCSR(csr *x509.CertificateRequest, earliestExpiry time.Time) (err error) {
+	parsedCertificate, err := x509.ParseCertificate([]byte(cert.DER))
+	if err != nil {
+		return
+	}
+
+	// Check issued certificate matches what was expected from the CSR
+	hostNames := make([]string, len(csr.DNSNames))
+	copy(hostNames, csr.DNSNames)
+	if len(csr.Subject.CommonName) > 0 {
+		hostNames = append(hostNames, csr.Subject.CommonName)
+	}
+	hostNames = UniqueNames(hostNames)
+
+	if !KeyDigestEquals(parsedCertificate.PublicKey, csr.PublicKey) {
+		err = InternalServerError("Generated certificate public key doesn't match CSR public key")
+		return
+	}
+	if len(csr.Subject.CommonName) > 0 && parsedCertificate.Subject.CommonName != csr.Subject.CommonName {
+		err = InternalServerError("Generated certificate CommonName doesn't match CSR CommonName")
+		return
+	}
```

Comment:
The other components of the name should be empty, e.g., `len(parsedCertificate.Subject.Country) == 0` should be true.  It would be good to test this for all the fields in `pkix.Name` besides `CommonName` (and `Names`).